lib/Doctrine/Common/Collections/Expr/ClosureExpressionVisitor.php

```diff
@@ -118,116 +114,68 @@ public static function sortByField(string $name, int $orientation = 1, ?Closure
-                return static function ($object) use ($field, $value): bool {
-                    return $value === substr(ClosureExpressionVisitor::getObjectFieldValue($object, $field), -strlen($value));
-                };
-
-            default:
-                throw new RuntimeException('Unknown comparison operator: ' . $comparison->getOperator());
-        }
+        $value = $comparison->getValue()->getValue();
+
+        return match ($comparison->getOperator()) {
+            Comparison::EQ => static fn ($object): bool => self::getObjectFieldValue($object, $field) === $value,
+            Comparison::NEQ => static fn ($object): bool => self::getObjectFieldValue($object, $field) !== $value,
+            Comparison::LT => static fn ($object): bool => self::getObjectFieldValue($object, $field) < $value,
+            Comparison::LTE => static fn ($object): bool => self::getObjectFieldValue($object, $field) <= $value,
+            Comparison::GT => static fn ($object): bool => self::getObjectFieldValue($object, $field) > $value,
+            Comparison::GTE => static fn ($object): bool => self::getObjectFieldValue($object, $field) >= $value,
+            Comparison::IN => static function ($object) use ($field, $value): bool {
+                $fieldValue = ClosureExpressionVisitor::getObjectFieldValue($object, $field);
+
+                return in_array($fieldValue, $value, is_scalar($fieldValue));
+            },
+            Comparison::NIN => static function ($object) use ($field, $value): bool {
+                $fieldValue = ClosureExpressionVisitor::getObjectFieldValue($object, $field);
+
+                return ! in_array($fieldValue, $value, is_scalar($fieldValue));
+            },
+            Comparison::CONTAINS => static fn ($object): bool => str_contains((string) self::getObjectFieldValue($object, $field), (string) $value),
+            Comparison::MEMBER_OF => static function ($object) use ($field, $value): bool {
+                $fieldValues = ClosureExpressionVisitor::getObjectFieldValue($object, $field);
+
+                if (! is_array($fieldValues)) {
+                    $fieldValues = iterator_to_array($fieldValues);
+                }
+
+                return in_array($value, $fieldValues, true);
+            },
+            Comparison::STARTS_WITH => static fn ($object): bool => str_starts_with((string) self::getObjectFieldValue($object, $field), (string) $value),
+            Comparison::ENDS_WITH => static fn ($object): bool => str_ends_with((string) self::getObjectFieldValue($object, $field), (string) $value),
+            default => throw new RuntimeException('Unknown comparison operator: ' . $comparison->getOperator()),
+        };
     }
 
     public function walkValue(Value $value): mixed
     {
         return $value->getValue();
     }
 
-    public function walkCompositeExpression(CompositeExpression $expr): mixed
+    public function walkCompositeExpression(CompositeExpression $expr): callable
```

Comment:
```suggestion
    public function walkCompositeExpression(CompositeExpression $expr): Closure
```

We could narrow the type here as well.